stdlib/LinearAlgebra/src/svd.jl

```diff
@@ -98,13 +98,23 @@ function svd!(A::StridedMatrix{T}; full::Bool = false, alg::Algorithm = default_
     end
     SVD(u,s,vt)
 end
+function svd!(A::StridedVector{T}; full::Bool = false, alg::Algorithm = default_svd_alg(A)) where {T<:BlasFloat}
+    if !full
+        normA = norm(A)
+        normalize!(A)
+        return SVD(reshape(A, (length(A), 1)), [normA], fill!(Matrix{T}(undef, 1, 1), 1))
```

Comment:
We need to reshape here to make the function type stable: in the other branch, the left factor is a matrix, so it must be here as well. Next, since the left factor must be of matrix type, so must be the right factor. Still, I observed a speed-up of almost factor 10 compared to calling the line below.

Reply:
I played around a little bit with using a simple Householder reflector to do the whole job for a vector. 
As basis I adapted the already existing code in `LinearAlgebra.reflector!`. Here is my code (still needs probably some polishing): 

```
function svd1!(A::StridedVector;full = false)
    # based on LinearAlgebra.reflector! 
    n = length(A)
    @inbounds begin
        ξ1 = A[1]
        normA = norm(A)
        if iszero(normA)
            full ? (return SVD(Matrix{eltype(A)}(I,n,n), [normA], ones(eltype(A),1,1) ) )  : 
                    (return SVD(Matrix{eltype(A)}(I,n,1), [normA], ones(eltype(A),1,1)) )
        end
        ν = copysign(normA, real(ξ1))
        ξ1 += ν
        A[1] = -ν
        for i = 2:n
            A[i] /= ξ1
        end
    end
    τ = ξ1/ν 
    A[1] = one(eltype(A))
    full ? (return SVD(I - (conj(τ)*A)*A', [normA], -ones(eltype(A),1,1)) )  : 
             (return SVD(Matrix{eltype(A)}(I,n,1)-conj(τ)*A, [normA], -ones(eltype(A),1,1)) )
end

n = 4
A = rand(n)
Asave = copy(A)
# full case
A = copy(Asave);
F = svd1!(A,full = true); F.U
F.U*F.S[1,1]*Matrix{eltype(A)}(I,4,1)*F.Vt - Asave
# thin case
A = copy(Asave);
F = svd1!(A,full = false); U1
F.U*F.S*F.Vt - Asave
```
With this occasion I observed that `svdvals` does not work on a vector:
```
svdvals([1])
ERROR: MethodError: no method matching svdvals(::Array{Int64,1})
Closest candidates are:
  svdvals(::LinearAlgebra.AbstractTriangular; kwargs...) at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\triangular.jl:2672
  svdvals(::StridedArray{T, 2}, ::StridedArray{T, 2}) where T<:Union{Complex{Float32}, Complex{Float64}, Float32, Float64} at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\svd.jl:510
  svdvals(::StridedArray{TA, 2}, ::StridedArray{TB, 2}) where {TA, TB} at D:\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v1.5\LinearAlgebra\src\svd.jl:536
  ...
Stacktrace:
 [1] top-level scope at REPL[193]:1
```

Reply:
> With this occasion I observed that `svdvals` does not work on a vector:

That's right, currently it doesn't, but it's included in this PR already.

Reply:
I have tested your reflector-based code, and it takes much much longer than the reshape-based version (as in e8285b4). I have also included your zero-norm case handling, thanks for the reminder!